### PR TITLE
Allow auto backups to Google sync

### LINKED ISSUE #109: Support Google Sync for backups (Where available)
This is a ticket to detail the requirements and current standing of the google sync support branch

As per https://developer.android.com/guide/topics/data/testingbackup.html

> If a device does not have any backup transports, then the data cannot be backed up. Your app is not adversely affected.

Non Google Play installs of the app will continue to function as they currently do, they just wont support this feature.

This will also only work if the user is using PIN/Password based encryption as the KeyStore based encryption is under the same limitations as with using something like Titanium Backup.

Lastly, this shouldnt be used as an alternative to the manual backups, I envision this to be used more in extenuating circumstances (phone breaks abroad and you dont have access to your backup) rather than as a backup alternative.

Current work is at https://github.com/RichyHBM/andOTP/tree/google-backups

### PR DESCRIPTION
**Pull request by RichyHBM**
Fixes #109 

This adds a backup agent which allows android to backup data onto the users google account, or any other backup-restore service installed onto their phone.

It will auto disable on changing back to keystore from password encryption

### CODE REVIEW

**Comment by flocke** [comment]
I'm gonna have a look at the implementation later on. But I will hold of the merge until after the release of 0.4.0 with the custom-encryption stuff.

**Comment by RichyHBM** [comment]
Makes total sense, I was under the impression it would be is own separate release anyway

File: app/src/main/java/org/shadowice/flocke/andotp/Utilities/BackupAgent.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public void onCreate() {
+        String prefs = getDefaultSharedPreferencesName();
+
+        SharedPreferencesBackupHelper sharedPreferencesBackupHelper = new SharedPreferencesBackupHelper(this, prefs);
+        addHelper(PREFS_BACKUP_KEY, sharedPreferencesBackupHelper);
+
+        FileBackupHelper fileBackupHelper = new FileBackupHelper(this, Constants.FILENAME_ENCRYPTED_KEY, Constants.FILENAME_DATABASE, Constants.FILENAME_DATABASE_BACKUP);
```
Comment (flocke) [inline app/src/main/java/org/shadowice/flocke/andotp/Utilities/BackupAgent.java]: You can drop the `Constants.FILENAME_ENCRYPTED_KEY` here. It is only used with the KeyStore encryption so it's no use backing it up.

**Comment by flocke** [comment]
I can't really test this since I don't use Google Backup (or any other automated device backup).
But I assume you tested it already @RichyHBM ? Maybe we should upload a test version again and let users try this out before releasing it to the public.

**Comment by RichyHBM** [comment]
I did yes, the android virtual emulators have a backup mechanism, I think by default they use google cloud but you can use adb to change them to use a debug backup service. I definitely didn't need to log in to a virtual device to test it.

Personally I think if anything arises from this it should just be a bug, it shouldn't delay this being in the release, but I'm happy to let others test it first

**Comment by flocke** [comment]
Alright that is totally fine with me!

File: app/src/main/res/values/strings_settings.xml [hunk 2]
```diff
@@ -63,6 +64,8 @@
     <string name="settings_desc_special_features">Uncheck to disable the special features again</string>
     <string name="settings_desc_enable_screenshot">Allow to take screenshots of the main screen
         (disabled by default for security reasons)</string>
+    <string name="settings_desc_enable_android_backup_service">Enables andOTP to use android\'s
+        built in backup service to bacup keys and preferences</string>
```
Comment (alexanderadam) [inline app/src/main/res/values/strings_settings.xml]: there's a small typo (`bac[k]up keys`)

Comment (flocke) [reply app/src/main/res/values/strings_settings.xml]: Just fixed it in 263a2097ce2ac8a71b4a41798dad5640df0f77a3, thanks for noticing.

Comment (alexanderadam) [reply app/src/main/res/values/strings_settings.xml]: Well, thank you for your work on andOTP :wink:

**Comment by 3xploiton3** [comment]
any update now?

**Pull request merged**